docs/features/communication/abi_compatible_data_types/index.rst

```diff
@@ -0,0 +1,491 @@
+
+Option Types
+""""""""""""
+
+ABI-compatible optional types must be implemented manually using a one-byte tag followed by a payload.
+
+.. code-block:: rust
+
+    #[repr(C)]
+    pub struct AbiOption<T> {
+        is_some: u8,
+        value: T,
+    }
+
+.. code-block:: cpp
+
+    template<typename T>
+    struct AbiOption {
+    private:
+        std::uint8_t is_some;
+        T value;
+    };
+
+* ``is_some == 0`` indicates absence; ``1`` indicates presence.
+* The value field is always initialized and occupies memory regardless of state.
+* The public API must match standard optional types in usability.
+
+Result Types
+""""""""""""
+
+Result types represent tagged unions with two possible states.
+
+.. code-block:: rust
+
+    #[repr(C)]
+    pub struct AbiResult<T, E> {
+        is_ok: u8,
+        value: AbiResultUnion<T, E>,
+    }
+
+    #[repr(C)]
+    union AbiResultUnion<T, E> {
+        ok: T,
+        err: E,
+    }
+
+.. code-block:: cpp
+
+    template<typename T, typename E>
+    struct AbiResult {
+    private:
+        std::uint8_t is_ok;
+        union {
+            T ok;
+            E err;
+        } value;
+    };
+
+* ``is_ok == 1`` indicates ``ok`` field is valid
+* ``is_ok == 0`` indicates ``err`` field is valid
+* The layout must guarantee correct union member interpretation based on the discriminant
+
+Language Conformance Summary
+""""""""""""""""""""""""""""
+
+.. list-table::
+   :header-rows: 1
+
+   * - Feature
+     - Rust Native Support
+     - C++ Native Support
+     - Specification Status
+   * - Primitives
+     - ✅ Native types
+     - ✅ Native types
+     - Conforming
+   * - Structs
+     - ✅ ``#[repr(C)]``
+     - ✅ ``standard_layout``
+     - Conforming
+   * - Enums (fieldless)
+     - ✅ ``#[repr(C)]``
+     - ✅ With fixed base
+     - Conforming
+   * - Arrays
+     - ✅ ``[T; N]``
+     - ✅ ``T[N]``
+     - Conforming
+   * - Vector
+     - ❌ ``Vec<T>``
+     - ❌ ``std::vector<T>``
+     - ✅ ``AbiVec<T, N>`` required
+   * - Option
+     - ❌ ``Option<T>``
+     - ❌ ``std::optional<T>``
+     - ✅ ``AbiOption<T>`` required
+   * - Result
+     - ❌ ``Result<T, E>``
+     - ❌ ``std::expected<T, E>``
+     - ✅ ``AbiResult<T, E>`` required
+
+
+
+Type Description
+----------------
+
+To address the scenarios outlined in the motivation, a clearly defined type description mechanism is required. The type description provides sufficient information during runtime, enabling a process without compile-time access to type definitions to correctly interpret a given memory location according to the previously established ABI rules.
+
+The goals are:
+
+* Enable interpretation of shared memory content without compile-time access to type definitions.
+* Support all ABI-compatible data types previously defined.
+* Include versioning to manage schema evolution and compatibility.
+* Allow easy generation and parsing by tooling in both C++ and Rust.
```

Comment:
Just as a side note, I personally think this is very hard to achieve with a type-safe, performant API and a static type system as in C++ and Rust.